File: sources/csharp/06_char_count/main.cs
// task 06 char_count — expected output: 10000000
// build: coreclr: dotnet build -c Release | nativeaot: dotnet publish -c Release -p:PublishAot=true | mono: mcs -optimize+ main.cs    run: coreclr: dotnet run -c Release (or bin/Release/net8.0/main.exe) | nativeaot: bin/Release/net8.0/publish/main.exe | mono: mono main.exe
using System;

class Program
{
    static void Main()
    {
        // Build the 100 MB text once, by repeating the whole 10-character block 10000000 times.
        string block = "abcdefghij";
        System.Text.StringBuilder builder = new System.Text.StringBuilder(100000000);
        for (int i = 0; i < 10000000; i++)
        {
            builder.Append(block);
        }
        string text = builder.ToString();

        long count = 0;
        for (int i = 0; i < text.Length; i++)
        {
            char ch = text[i];
            if (ch == 'h')
            {
                count++;
            }
        }
        Console.WriteLine(count);
    }
}

File: sources/c/06_char_count/main.c
// task 06 char_count — expected output: 10000000
// build: gcc -O2 -pthread -o prog main.c    run: ./prog
// alternates: clang -O2 -pthread -o prog main.c | cl /O2 /Fe:prog main.c | tcc -o prog main.c

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#define BLOCK_LEN 10
#define REPEATS 10000000LL
#define TEXT_LEN (REPEATS * BLOCK_LEN)

int main(void) {
    char *text = (char *)malloc((size_t)TEXT_LEN);

    /* the whole 100 MB text is built up front, block by block */
    for (int64_t i = 0; i < REPEATS; i++) {
        memcpy(text + i * BLOCK_LEN, "abcdefghij", BLOCK_LEN);
    }

    int64_t count = 0;
    for (int64_t i = 0; i < TEXT_LEN; i++) {
        if (text[i] == 'h') {
            count += 1;
        }
    }

    printf("%lld\n", (long long)count);
    free(text);
    return 0;
}
